Apply ZIO SBT Ecosystem Plugin

khajavi opened the pull request.

khajavi committed:
initial work.

khajavi committed:
zio config settings.

khajavi committed:
js, jvm and native settings.

khajavi committed:
zioConfigAws.

khajavi committed:
enable streaming.

khajavi committed:
zioConfigAwsJVM.

khajavi committed:
zioConfigAws.

khajavi committed:
more setting migration.

khajavi committed:
more settings.

khajavi committed:
move most settings to new version.

khajavi committed:
remove scala211.

khajavi committed:
remove build helper.

khajavi committed:
scala 211.

khajavi committed:
stop forcing unsupported scala version.

khajavi committed:
remove website workflow.

khajavi committed:
update scala versions.

khajavi committed:
remove scala211 from zio config js module.

khajavi committed:
prevent turning warnings to errors.

khajavi committed:
generate ci.

khajavi committed:
remove scala211.

khajavi committed:
nowarn.

khajavi committed:
update enumeratum.

khajavi committed:
remove noindent scalac options.

khajavi committed:
make the root cross scala version empty.

khajavi committed:
update zio-sbt plugins.

khajavi committed:
remove scala3 from zioConfigPureJVM module.

khajavi committed:
remove scala3 from zioConfigTypesafeMagnoliaTestsJVM module.

khajavi committed:
regenerate ci workflow.

khajavi committed:
Revert "remove scala3 from zioConfigPureJVM module."

This reverts commit 3501ad53d6d6851d5cda9ad13960b097508ac916.

khajavi committed:
fmt.

khajavi committed:
fmt and fix.

khajavi committed:
remove extra command aliases.

khajavi committed:
update ci.

khajavi committed:
remove scala3 from two modules.

khajavi committed:
generate ci.

khajavi commented:
This work is ready for review.

The failing CI seems related to this issue #1095

afsalthaj commented:
Thanks Milad. Have been waiting for this plugin and great addition :-)

afsalthaj commented on .github/workflows/ci.yml:
```diff
@@ -1,89 +1,215 @@
-name: CI
+# This file was autogenerated using `zio-sbt-ci` plugin via `sbt generateGithubWorkflow` 
+# task and should be included in the git repository. Please do not edit it manually.
```
👍

afsalthaj commented on .scalafmt.conf:
```diff
@@ -1,16 +1,19 @@
-  "core/src/test/scala/zio/config/ProductBuilderTest.scala"
+  "core/src/test/scala/zio/config/ProductBuilderTest.scala",
+  "core/shared/src/main/scala-3.x/*",
+  "magnolia/shared/src/main/scala-dotty/*"
```
I think this can also go into a custom module within ecosystem plugin that wraps scalafmt
And for custom settings, such as exclude, I could simply configure with in each module setting, `excludeFmt`

khajavi replied:
Thank you! Added to the backlog.

afsalthaj commented on build.sbt:
```diff
@@ -1,11 +1,13 @@
-    developers := List(
+    name                       := "ZIO Config",
+    crossScalaVersions -= scala211.value,
+    ciEnabledBranches          := Seq("series/4.x"),
```
I think we should include `master` as well, since `master` works with latest zio

khajavi replied:
What about series/2.x? Doesn't it work for the latest version of ZIO?

afsalthaj replied:
I don't think we need to support 2.x going forward

khajavi replied:
Done!

afsalthaj commented on build.sbt:
```diff
@@ -1,11 +1,13 @@
-    licenses := List("Apache-2.0" -> url("http://www.apache.org/licenses/LICENSE-2.0")),
-    developers := List(
+    name                       := "ZIO Config",
+    crossScalaVersions -= scala211.value,
```
I would rather just rely on ecosystem plugin to decide the common scala versions that an ecosystem library should support. Obviously, ecosystem libraries can choose to exclude from this common list. This implies there will not be specific `inThisBuild` settings for crossScalaVersions. The exclusion, should be within each module.
Ecosystem Plugin can also choose to pre-built excluded list.

khajavi replied:
Currently, zio-sbt provides default scala versions, then we can add/remove and customize that for the whole modules or any submodule. 
I'm not sure I got what you said, here I excluded scala 211 from all modules, although we can do that one by one.

afsalthaj commented on build.sbt:
```diff
@@ -18,31 +20,30 @@ inThisBuild(
 
-addCommandAlias("fmt", "; scalafmtSbt; scalafmt; test:scalafmt")
-addCommandAlias("fix", "; all compile:scalafix test:scalafix; all scalafmtSbt scalafmtAll")
 addCommandAlias("compileAll", "; ++2.12.16; root2-12/compile; ++2.13.8!; root2-13/compile")
```
I would delete this as well. CommandAlias can be a pain. I would either choose it as sbtSettings  that are not based on String. Even better is within ecosystem plugin

khajavi replied:
Done!

afsalthaj commented on build.sbt:
```diff
@@ -18,31 +20,30 @@ inThisBuild(
-addCommandAlias("fmt", "; scalafmtSbt; scalafmt; test:scalafmt")
-addCommandAlias("fix", "; all compile:scalafix test:scalafix; all scalafmtSbt scalafmtAll")
 addCommandAlias("compileAll", "; ++2.12.16; root2-12/compile; ++2.13.8!; root2-13/compile")
 addCommandAlias("testAll", "; ++2.12.16; root2-12/test; ++2.13.8!; root2-13/test")
```
I would rather choose the same Same as above

khajavi replied:
Done!

afsalthaj commented on build.sbt:
```diff
@@ -65,13 +67,14 @@ lazy val magnoliaDependencies =
 lazy val refinedDependencies =
   libraryDependencies ++= {
-    if (scalaBinaryVersion.value == "2.11") Seq.empty // Just to make IntelliJ happy
+    if (scalaVersion.value == scala211.value) Seq.empty // Just to make IntelliJ happy
```
This can happen several times. One setting can be `addDependency` which takes simply takes a list of scala versions to be excluded. Otherwise, different projects may have different ways to handle dependency settings. 

Also `refinedDependencies` is an unnecessary redirection, that it could be inlined as well

khajavi replied:
That's a good idea, but it is the first time I saw such a case. I'm doing the same for other ecosystem projects. If I saw that the common case, I'll make a general utility inside zio-sbt.

khajavi replied:
Just wrote these utility functions:

```scala
def addDependenciesFor(scalaBinaryVersions: String*)(dependencies: ModuleID*) =
  libraryDependencies ++= {
    if (scalaBinaryVersions.contains(scalaBinaryVersion.value)) dependencies else Seq.empty
  }

def addDependenciesExceptFor(scalaBinaryVersions: String*)(dependencies: ModuleID*) =
  libraryDependencies ++= {
    if (scalaBinaryVersions.contains(scalaBinaryVersion.value)) Seq.empty else dependencies
  }

def addScalacOptionsFor(scalaBinaryVersions: String*)(options: String*) =
  scalacOptions ++= {
    if (scalaBinaryVersions.contains(scalaBinaryVersion.value)) options else Seq.empty
  }

def addScalacOptionsExceptFor(scalaBinaryVersions: String*)(options: String*) =
  scalacOptions ++= {
    if (scalaBinaryVersions.contains(scalaBinaryVersion.value)) Seq.empty else options
  }
```

afsalthaj commented on build.sbt:
```diff
@@ -138,116 +141,91 @@ lazy val `root3` =
-    ),
-    testFrameworks := Seq(new TestFramework("zio.test.sbt.ZTestFramework"))
+      "com.amazonaws" % "aws-java-sdk-ssm" % awsVersion
+    )
```
Much better for any new contributor, is to be able to say

```scala
lazy val zioConfigAws = jvmModule("aws").withDependencies().dependOn(zioConfig)
```

As mentioned above withDependencies can take care of weird cases, such as excluding if Scala-version is `x` etc.

The ecosystem plugin mandates the name of the root project (say zio-config), and any module definition simply append the file path name. This way every project becomes a standard

khajavi replied:
I agree, will work on this!

afsalthaj commented on build.sbt:
```diff
@@ -272,176 +250,139 @@ lazy val examplesJVM = examples.jvm
-  .settings(stdSettings("zio-config-scalaz"))
-  .settings(crossProjectSettings)
+  .settings(stdSettings(name = "zio-config-scalaz", enableCrossProject = true))
+  .settings(enableZIO())
```
I think its mostly to help defining modules easily without relying on adhoc sbt settings. As mentioned above, as a maintainer I am far relieved if I can just simply define `module(...)` using the plugin, or if needed `jvmModule`, `jsModule` etc.

afsalthaj commented:
I have put some comments, but all of it boils down to 1 suggestion, that being able to define module based on the plugin, 

```scala
lazy val myModule = 
   module("aws")
    .withDependencies("dev.zio" %%% "izumi-reflect" % "2.2.2")
    .dependsOn(core)
```

Adding customisations becomes even simpler

```scala
lazy val myModule = 
   module("aws")
    .skipScala211
    .withDependencies("dev.zio" %%% "izumi-reflect" % "2.2.2")
    .dependsOn(core)

```

This is much easier for a maintainer. Example: In the above case, `core` can support 211, but `aws` cannot, and there is no need of an indirection such as having a big Map of each module and scala versions supported

khajavi commented:
@afsalthaj 
That is an excellent DSL for writing builds without any boilerplate. I agree with such language. The first iteration aimed to factor out `BuildHelper` and make it maintainable. 
I am excited to work on implementing your suggestions for the next phase.

khajavi commented:
> In the above case, `core` can support 211, but `aws` cannot, and there is no need of an indirection such as having a big Map of each module and scala versions supported

The name of `supportedScalaVersions` is somehow confusing, I will rename it to something like `ciScalaVersionMatrix`. We just use that to create the test matrix. This is just a **workaround** because I lack deep knowledge of how to iterate through sbt submodules and create matrix map pragmatically.

So ideally I would like to have something like this:

```scala
val ciScalaVersionsMatrix: Map[String, Seq[String]] = 
  listOfAllModules.map { module =>
      (module / thisProject).value.id   -> (module / crossScalaVersions).value
  }
```

But it will not compile due to complicated macro-based sbt internals. We need to find sbt specific solution to traverse modules and pull out settings.

khajavi committed:
remove extra commands.

khajavi committed:
update ci.

khajavi committed:
include master branch for ci trigger.

khajavi committed:
increase initial and maximum heap size.

khajavi committed:
set max parallel test to 4.

we should monitor and investigate to find the best value.

khajavi committed:
set max parallel test to 8.

with 4, it takes about one hour and 14 minutes to run all jobs.

khajavi committed:
scalafix.

khajavi committed:
fix docs generation warning.

khajavi committed:
fmt.

khajavi committed:
scalafix.

khajavi committed:
add back scala 3 support for zioConfigMagnoliaJVM and zioConfigRefinedJVM

khajavi commented:
Thrilled to see the geeeen color 😅
@afsalthaj Can you please have a look?

khajavi committed:
fmt

khajavi committed:
fmt.

khajavi committed:
update zio-sbt plugins.

khajavi committed:
add utility functions for add dependencies and scalac options.

khajavi committed:
increase Xmax-inlines to 64.

Error message:
[error] -- Error: /home/runner/work/zio-config/zio-config/magnolia/shared/src/test/scala-dotty/zio/config/magnolia/MarkdownSpec.scala:14:44
[error]  14 |        generateDocs(deriveConfig[RawConfig]).toTable.toGithubFlavouredMarkdown
[error]     |                                            ^
[error]     |   given instance derived is declared as `inline`, but was not inlined
[error]     |
[error]     |   Try increasing `-Xmax-inlines` above 32

khajavi committed:
fix name for docs project.

khajavi committed:
fix build errors.

khajavi committed:
remove scala 3 support from refined module.

To reproduce the following error, run `sbt zioConfigRefinedJVM/Test/compile`
[error] -- [E008] Not Found Error: /home/runner/work/zio-config/zio-config/refined/shared/src/test/scala/zio/config/refined/NumericSupportTest.scala:3:26
[error] 3 |import eu.timepit.refined.W
[error]   |                          ^
[error]   |                          value W is not a member of eu.timepit.refined

khajavi committed:
update zio-sbt plugins.

khajavi committed:
scalafix.

khajavi committed:
remove CheckAllCodeCompiles step from ci.

khajavi committed:
refined: add support for scala 3.

afsalthaj committed:
Merge branch 'series/4.x' into apply-zio-ecosystem-plugin

khajavi committed:
Merge branch 'series/4.x' into apply-zio-ecosystem-plugin
